File: 2015/swift/3/main-part-2.swift
import Foundation

protocol Moveable {
    mutating func move(direction: Character)
}

struct Position: Equatable {
    var x: Int
    var y: Int

    static func == (lhs: Position, rhs: Position) -> Bool {
        return lhs.x == rhs.x && lhs.y == rhs.y
    }
}

extension Position: Moveable {
    mutating func move(direction: Character) {
        switch direction {
        case "^":
            self.y += 1
        case "v":
            self.y -= 1
        case "<":
            self.x -= 1
        case ">":
            self.x += 1
        default:
            break
        }
    }
}

func main() {
    let arguments = CommandLine.arguments

    if arguments.count < 2 {
        print("Usage: swift main.swift <input_file>")
        return
    }

    let inputFilePath = arguments[1]
    let input = try! String(contentsOfFile: inputFilePath)
    let lines = input.split(separator: "\n")

    var santa = Position(x: 0, y: 0)
    var robo_santa = Position(x: 0, y: 0)

    // hashmap of visited positions
    var visited = [Position]()
    visited.append(santa) 

    var santa_turn = true

    for line in lines {
        for char in line {
            if santa_turn {
                santa.move(direction: char)
                if !visited.contains(santa) {
                    visited.append(santa)
                }
            } else {
                robo_santa.move(direction: char)
                if !visited.contains(robo_santa) {
                    visited.append(robo_santa)
                }
            }
            santa_turn = !santa_turn
        }
    }

    print("Part 2: \(visited.count)")
}

main()

File: 2015/javascript/3/main-part-2.js
class Position {
    constructor(x, y){
        this.x = x;
        this.y = y;
    }

    move(direction){
        switch(direction){
            case '^':
                this.y++;
                break;
            case 'v':
                this.y--;
                break;
            case '>':
                this.x++;
                break;
            case '<':
                this.x--;
                break;
        }
    }
}

function main(){
    // get command line arguments
    const args = process.argv.slice(1);

    if(args.length < 1){
        console.log("Usage: node main.js <input_file>");
        return;
    }

    // get input file
    const filename = args[1];
    const fs = require('fs');
    if(!fs.existsSync(filename)){
        console.log(`File ${filename} not found`);
        return;
    }

    // read input file
    const data = fs.readFileSync(filename, 'utf8');
    const lines = data.split('\n');

    let santa = new Position(0, 0);
    let robo_santa = new Position(0, 0);
    let santa_turn = true;

    // create 2d hashmap to store visited positions
    let grid = {};
    grid[santa.x] = {};
    grid[santa.x][santa.y] = 1;
    grid[robo_santa.x][robo_santa.y] = 1;

    for(let i = 0; i < lines.length; i++){
        const line = lines[i];
        for(let j = 0; j < line.length; j++){
            const char = line[j];
            let position = santa_turn ? santa : robo_santa;
            position.move(char);
            if (grid[position.x] === undefined){
                grid[position.x] = {};
                grid[position.x][position.y] = 1;
            }
            else if (grid[position.x][position.y] === undefined){
                grid[position.x][position.y] = 1;
            }
            else {
                grid[position.x][position.y]++;
            }
            santa_turn = !santa_turn;
            
        }
    }

    let count = 0;
    for (let x in grid){
        for (let _ in grid[x]){
            count++;
        }
    }


    console.log("Part 2: " + count);
}

main();